File: codes/Garnetwzy/20.cpp
class Solution {
public:
   bool isValid(string s) {
    stack<char> st;
    char p;
    for(int i=0;s[i]!='\0';i++){
        if(st.empty()){
            if (s[i]!=')'&&s[i]!=']'&&s[i]!='}') {
                st.push(s[i]);
            }else{
                return false;
            }
        }else{
            switch (s[i]){
                case '(':
                case '[':
                case '{':
                    st.push(s[i]);
                    break;
                case ')':
                    if(st.top()=='('){
                        st.pop();
                    }else{
                        return false;
                    }
                    break;
                case ']':
                    if(st.top()=='['){
                        st.pop();
                    }else{
                        return false;
                    }
                    break;
                case '}':
                    if(st.top()=='{'){
                        st.pop();
                    }else{
                        return false;
                    }
                    break;
                default:
                    break;
            }
        }
        
    }
    if(st.empty()){
        return true;
    }
    return false;
}
};

File: codes/HellDream/20.java
class Solution {
    public boolean isValid(String s) {
        if(s==null || s.equals("")) return true;
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> map  =new HashMap<>();
        map.put(')','(');
        map.put(']','[');
        map.put('}','{');
        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);
            if(c == '('||c == '['||c=='{'){
                stack.push(c);
            }else{
                if(!stack.isEmpty()&&stack.peek()==map.get(c)){
                    stack.pop();
                }else{
                    return false;
                }
            }
        }
        if(stack.isEmpty()) return true;
        return false;
    }
}